api/app/Models/PoolCandidate.php

```diff
@@ -911,4 +919,167 @@ public function setApplicationSnapshot()
+            }
+
+            if ($hasEssentialSkillsToAssess) {
+
+                // Check assessed essential skills on this step
+                $essentialSkillAssessments = $stepResults
+                    ->filter(function ($result) {
+                        return $result->poolSkill?->type === PoolSkillType::ESSENTIAL->name &&
+                            $result->assessment_result_type !== AssessmentResultType::EDUCATION;
+                    });
+
+                if ($essentialSkillAssessments->isEmpty()) {
+                    $hasToAssess = true;
+                }
+
+                foreach ($essentialSkillAssessments as $essentialSkillAssessment) {
+                    $decision = $essentialSkillAssessment->assessment_decision;
+
+                    switch ($decision) {
+                        case null:
+                            $hasToAssess = true;
+                            break;
+                        case AssessmentDecision::HOLD->name:
+                            $hasOnHold = true;
+                            break;
+                        case AssessmentDecision::UNSUCCESSFUL->name:
+                            $hasFailure = true;
+                            break;
+                        default:
+                    }
+                }
+            }
+
+            // Check for education requirement if is application screening step
+            if ($step->type === AssessmentStepType::APPLICATION_SCREENING->name) {
+                $educationResults = $stepResults->where('assessment_result_type', AssessmentResultType::EDUCATION->name);
+
+                if (! $educationResults) {
+                    $hasToAssess = true;
+                }
+
+                foreach ($educationResults as $educationResult) {
+                    $decision = $educationResult->assessment_decision;
+                    switch ($decision) {
+                        case null:
+                            $hasToAssess = true;
+                            break;
+                        case AssessmentDecision::HOLD->name:
+                            $hasOnHold = true;
+                            break;
+                        case AssessmentDecision::UNSUCCESSFUL->name:
+                            $hasFailure = true;
+                            break;
+                        default:
+                    }
+                }
+            }
+
+            if ($hasFailure) {
+                $decisions[] = [
+                    'step' => $stepId,
+                    'decision' => AssessmentDecision::UNSUCCESSFUL->name,
+                ];
+
+                continue;
+            }
+
+            if ($hasToAssess) {
+                $decisions[] = [
+                    'step' => $stepId,
+                    'decision' => null,
+                ];
+
+                continue;
+            }
+
+            // Candidate has been assessed and was not unsuccessful so continue to next step
+            $currentStep++;
```

Comment:
I just realized, I think this will cause an incorrect currentStep in some cases.

Add a phpunit test for this case: There are two assessment steps. The first one is unassessed (or even unsuccessful). The second step is fully assessed, successfully. `currentStep` should be locked to step one. 

I'm worried that with the current implementation, the later steps in the loop will increment the `currentStep` past where it should get stuck.

Reply:
You are correct. Added some logic to only increment when previous steps have been assessed and not been marked as unsuccessful

[do not increment step when previous steps not passed](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10642/commits/943049c819e067eddbd8a9a6a07310c5ca030780)